freqtrade/freqai/data_kitchen.py

```diff
@@ -0,0 +1,1215 @@
+            dataframe = prediction_dataframe.copy()
+            for tf in tfs:
+                base_dataframes[tf] = None
+                for p in pairs:
+                    if p not in corr_dataframes:
+                        corr_dataframes[p] = {}
+                    corr_dataframes[p][tf] = None
+        else:
+            dataframe = base_dataframes[self.config["timeframe"]].copy()
+
+        sgi = False
+        for tf in tfs:
+            if tf == tfs[-1]:
+                sgi = True  # doing this last allows user to use all tf raw prices in labels
+            dataframe = strategy.populate_any_indicators(
+                pair,
+                dataframe.copy(),
+                tf,
+                informative=base_dataframes[tf],
+                set_generalized_indicators=sgi
+            )
+            if pairs:
+                for i in pairs:
+                    if pair in i:
+                        continue  # dont repeat anything from whitelist
+                    dataframe = strategy.populate_any_indicators(
+                        i,
+                        dataframe.copy(),
+                        tf,
+                        informative=corr_dataframes[i][tf]
+                    )
+
+        return dataframe
+
+    def fit_labels(self) -> None:
+        """
+        Fit the labels with a gaussian distribution
+        """
+        import scipy as spy
+
+        self.data["labels_mean"], self.data["labels_std"] = {}, {}
+        for label in self.data_dictionary["train_labels"].columns:
+            if self.data_dictionary["train_labels"][label].dtype == object:
+                continue
+            f = spy.stats.norm.fit(self.data_dictionary["train_labels"][label])
+            self.data["labels_mean"][label], self.data["labels_std"][label] = f[0], f[1]
+
+        # KEEPME incase we want to let user start to grab quantiles.
+        # upper_q = spy.stats.norm.ppf(self.freqai_config['feature_parameters'][
+        #                                                   'target_quantile'], *f)
+        # lower_q = spy.stats.norm.ppf(1 - self.freqai_config['feature_parameters'][
+        #                                                       'target_quantile'], *f)
+        # self.data["upper_quantile"] = upper_q
+        # self.data["lower_quantile"] = lower_q
+        return
+
+    def remove_features_from_df(self, dataframe: DataFrame) -> DataFrame:
+        """
+        Remove the features from the dataframe before returning it to strategy. This keeps it
+        compact for Frequi purposes.
+        """
+        to_keep = [
+            col for col in dataframe.columns if not col.startswith("%") or col.startswith("%%")
+        ]
+        return dataframe[to_keep]
+
+    def get_current_trade_database(self) -> None:
+
+        if self.database_path is None:
+            logger.warning('No trade database found. Skipping analysis.')
+            return
+
+        data = sqlite3.connect(self.database_name)
```

Comment:
Freqtrade does support other database engines as well (postgres, mariadb).
using sqlite3 will therefore fail in such instances (this is therefore introducing a bug for such instances - especially as no check for "is this a sqlite url" is in place).
so you either need to account for these instances (careful as psycopg2 and pymysql are not required by default, so imports may fail) - or handle this differently (which is what i'd suggest).

You should be able to use something around `Trade._session.bind` - which you can pass to `pd.from_sql()` - and which will reuse the existing connection.
Obviously it'll not work for backtesting - but that'll not work with a database anyway i'd think.

Reply:
@robcaulk i don't really see this as resolved if i'm honest.

It's still not supporting other engines - and it's also not reusing available methods (there's a method to load trade data from DB - as well as a (threaded) session that exists for the lifetime of the bot.

Assuming you see this as temporary workaround (a workaround i clearly don't like):
Also error handling around this is clearly not sufficient - if something between `.connect()` and `.close()` fails - you'll keep the session open, potentially causing a lock on the database (sqlite is pretty prone to locking problems ...).